core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/DetermineJoinDistributionType.java

```diff
@@ -105,20 +111,92 @@ public static double getSourceTablesSizeInBytes(PlanNode node, Context context)
 
     @VisibleForTesting
     static double getSourceTablesSizeInBytes(PlanNode node, Lookup lookup, StatsProvider statsProvider, TypeProvider typeProvider)
+    {
+        List<PlanNode> sourceNodes = getSourceNodes(node, lookup);
+        if (sourceNodes.isEmpty()) {
+            return NaN;
+        }
+
+        return sourceNodes.stream()
+                .mapToDouble(sourceNode -> statsProvider.getStats(sourceNode).getOutputSizeInBytes(sourceNode.getOutputSymbols(), typeProvider))
+                .sum();
+    }
+
+    private static double getApproximateSourceSizeInBytes(PlanNode node, Context context)
+    {
+        return getApproximateSourceSizeInBytes(node, context.getLookup(), context.getStatsProvider(), context.getSymbolAllocator().getTypes());
+    }
+
+    @VisibleForTesting
+    static double getApproximateSourceSizeInBytes(PlanNode node, Lookup lookup, StatsProvider statsProvider, TypeProvider typeProvider)
```

Comment:
I don't think `getApproximateSourceSizeInBytes` properly describes what this method is doing (because of row filtering factor multiplication). Maybe:
```
getApproximateOutputSizeInBytes
```
or
```
getFirstKnownOutputSizeInBytes
```
if we simplify approach